### PR TITLE
Improve IPO-safety of TBAA information 

### PR DESCRIPTION
**Pull request by Keno**
I've been experimenting with enabling some IPO passes on our generated code.
Unfortunately at the moment, we're seeing some pretty major miscompiles from
code like the following:

```
define @foo() {
%local = alloca
store %local, !tbaa tbaa_stack
call @bar(%local)
}
define @bar(%arg) {
load %arg, !tbaa tbaa_const
}
```

Now, in the absence of IPO, this IR is mostly fine. We don't modify %arg during
the course of bar and don't otherwise reference it. However, once IPO comes into
play, the above IR is actually illegal, because the %local in foo and the
%arg in %bar alias, but have different TBAA information. Because of this,
LLVM can assume that they don't alias (even if it's other alias analyses tell
it that they must alias) and will happily replace the load value with undef
to punish us for our subversive behavior.

LLVM is not entirely incorrect here. The constant flag in TBAA is meant to apply
only to data that is globally constant (e.g. because it was initialized by the dynamic
library loader).

To please LLVM here, we need to stop annotating tbaa_const on function arguments.
However, doing so without coming up with some replacement would be quite bad for performance,
because LLVM would no longer be able to optimize these values as aggressively. So here's the
replacement:
```
define @bar(%arg) {
%arg2 = @llvm.invariant.group.barrier(%arg)
load %arg2, !invariant.group !0
}
```
The invariant group allows LLVM to assume that any load from that pointer with the annotated
inariant group loads the same value, so in particular, it's allowed to fold:
```
define @bar(%arg) {
%arg2 = @llvm.invariant.group.barrier(%arg)
load %arg2, !invariant.group !0
call @external()
load %arg2, !invariant.group !0
}
```
by removing the second load.

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -5593,7 +5611,9 @@ static std::unique_ptr<Module> emit_function(
                     theArg = ghostValue(argType);
                 }
                 else if (llvmArgType->isAggregateType()) {
-                    theArg = mark_julia_slot(&*AI++, argType, NULL, tbaa_const); // this argument is by-pointer
+	            Argument *Arg = &*AI++;
```
Comment (tkelman) [inline src/codegen.cpp]: there's a tab here that should be spaces

**Comment by Keno** [comment]
Do we run something older than 3.9.1 on CI? I don't see the failure to select locally.

**Comment by tkelman** [comment]
It might not have all the very latest patches on mac or windows, would need to rebuild the binaries there

**Comment by Keno** [comment]
Linux fails as well. I'll rebuild 3.9.1 with assertions (I only locally have svn with assertions and 3.9.1 without).

**Comment by Keno** [comment]
So looks like there was https://reviews.llvm.org/D26292 which fixed this. Don't think it's worth carrying the patch. Let's upgrade to 4.0 shortly after 0.6 is out at which point we can merge this.

**Comment by tkelman** [comment]
that plan's alright with me, if there's no hurry on this

**Comment by Keno** [comment]
We may want to consider dropping the tbaa_const anyway for correctness, even if we can't apply the invariant_group change, but I'm somewhat worried about performance regressions.

**Comment by vtjnash** [comment]
> for correctness

Our current IR is correct, as long as you don't enable LTO and change the scope to which the tbaa information is getting applied. But it would be good to get the array owner write barrier fix in a separate PR.

**Comment by Keno** [comment]
I ended up determining this use of invariant group was incorrect, but was able to achieve the same performance with more precise TBAA, which I'll put up soon.

**Pull request closed without merging**